---  
  
Name : Nicku Rendy Perdana  
Batch : HCK-012

Program ini diperuntukkan sebagai kisi-kisi Live Code 2 untuk Phase 0 Fulltime Data Science Hacktiv8 Kampus Pondok Indah

---

# **Kisi-kisi LC2**

Dataset yang digunakan pada latihan ini berasal dari bigquery public data. Lakukan koneksi Google Colaboratory dengan BigQuery menggunakan syntax berikut:

```
from google.colab import auth
from google.cloud import bigquery
auth.authenticate_user()
print('Authenticated')

project_id = "hckv8-95102" #GUNAKAN GCP PROJECT-ID KALIAN MASING-MASING
client = bigquery.Client(project=project_id)
```

lalu lakukan query pada notebook menggunakan cara berikut:

```
df_1 = client.query('''
SELECT *
FROM table_name`
''').to_dataframe()

df_1
```

---

Dataset yang digunakan merupakan `record perjalanan peminjam sepeda`, dapat diakses pada:

- **Project_id_Database**: hacktiv8-ftds-challenge
- **Dataset**            : practice
- **Table**              : bike_rent

Lakukan pengolahan SQL langsung melalui Google Colaboratory yang telah dikoneksikan pada BigQuery

---
Pada notebook ini terdapat 2 section, **SQL** dan **Pandas**.

Kerjakan masing-masing section sesuai dengan ketentuan di bawah:

## A. **SQL**

  >**Jawab masing-masing soal dalam 1 query.**

  1. Berapa banyak label unique dari `bike_type`?  
  2. Apa saja nama label unique dari tipe sepeda?
  3. Berapa banyak total peminjaman sepeda pada tahun 2016 dan 2017?
  4. Dimana station yang paling banyak dikunjungi sebagai titik keberangkatan pada tahun 2016?
  5. Berapa nilai minimum, maximum, dan rata-rata dari durasi perjalanan pada keseluruhan data?
  6. Berapa rata-rata durasi perjalanan dari subscriber_type `Explorer` dan `Weekender`?
  7. Tampilkan entry data yang memiliki `duration_time` diatas 4.25 jam. (*Gunakan subquery untuk menjawab pertanyaan ini dengan meng-convert duration_time dari menit ke hour*).

## B. **Pandas**
  >**Data yang akan diolah menggunakan Pandas wajib diambil menggunakan Query berikut:**
  ```
  SELECT *
  FROM `hacktiv8-ftds-challenge.practice.bike_rent`
  ```
  
  >Simpan data yang telah diambil ke dalam variabel `df` (penamaan variabel dibebaskan), **lakukan pengolah di bawah ini menggunakan library Pandas**:

  1. Cek kesesuaian tipe data dan keberadaan missing values dengan `.info()`

  2. Buat kolom baru bernama `quarter`.
  
  Lakukan ekstraksi waktu (quarter dalam tahun) berdasarkan kolom `start_time` dengan ketentuan berikut:

<div align="center">

  |Keterangan Waktu|Bulan|
  |--|--|
  |Q1|Januari, Februari, Maret|
  |Q2|April, Mei, Juni|
  |Q3|Juli, Agustus, September|
  |Q4|Oktober, November, Desember|

</div>

  3. Tampilkan data perjalanan dari tipe member `Local365` dengan durasi perjalanan diatas 1 jam *`(Jika tidak ada entry yang memenuhi kondisi ini tidak apa, tuliskan kode pandasnya dengan lengkap saja)`*.

  4. Berapa rata-rata durasi perjalanan dari tipe pelanggan `Walk Up` dengan tipe sepeda `classic` pada bulan `Maret` berdasarkan data yang diambil? *`(Jika tidak ada entry yang memenuhi kondisi ini tidak apa, tuliskan kode pandasnya dengan lengkap saja)`*.

>**Notes:**
>- Dalam menjawab soal Pandas no. 2, **tidak boleh** menggunakan ekstraksi waktu `Series.dt.quarter`
- Lakukan `looping` dan `conditional` dengan memanfaatkan ekstraksi bulan `Series.dt.month` terhadap kolom `start_time`, pastikan tipe datanya ialah `datetime` agar dapat dilakukan ekstraksi.
- Simpan hasil looping ke dalam suatu variabel, dan buat kolom baru berisikan variabel sebelumnya.



###**SQL**

In [1]:
from google.colab import auth
from google.cloud import bigquery
import pandas as pd

auth.authenticate_user()

project_id = "data-science-408704"
client = bigquery.Client(project=project_id)

In [2]:
df = client.query('''
SELECT *
FROM `hacktiv8-ftds-challenge.practice.bike_rent`
''').to_dataframe()

df

trip_id subscriber_type bike_id bike_type                start_time  \
0     18896589        Explorer      93   classic 2018-10-28 14:16:14+00:00   
1     13948351        Local365     315   classic 2017-03-07 12:37:33+00:00   
2     16243803        Local365     877   classic 2017-10-21 09:49:48+00:00   
3     16222446         Walk Up     273   classic 2017-10-18 15:01:52+00:00   
4     16222451         Walk Up     201   classic 2017-10-18 15:02:26+00:00   
...        ...             ...     ...       ...                       ...   
7995  15469480         Walk Up     327   classic 2017-07-22 20:17:46+00:00   
7996  16130650         Walk Up     528   classic 2017-10-07 11:25:36+00:00   
7997  14204138         Walk Up     235   classic 2017-03-27 13:19:02+00:00   
7998  15037809         Walk Up     206   classic 2017-06-05 11:42:40+00:00   
7999   9499076         Walk Up     315   classic 2016-04-09 18:52:09+00:00   

                           start_station_name  \
0                     Dean Keeton & Speedway    
1                                 Main Office   
2           Republic Square @ 5th & Guadalupe   
3             Bullock Museum @ Congress & MLK   
4             Bullock Museum @ Congress & MLK   
...                                       ...   
7995          Bullock Museum @ Congress & MLK   
7996                  Congress & Cesar Chavez   
7997          State Capitol @ 14th & Colorado   
7998          State Capitol @ 14th & Colorado   
7999  Capital Metro HQ - East 5th at Broadway   

                             end_station_name  duration_minutes  
0                              11th & Salina                178  
1                                 Main Office               198  
2                                 Main Office                60  
3                                 Main Office                60  
4                                 Main Office                60  
...                                       ...               ...  
7995          Bullock Museum @ Congress & MLK                55  
7996          Bullock Museum @ Congress & MLK                55  
7997          State Capitol @ 14th & Colorado                55  
7998          State Capitol @ 14th & Colorado                55  
7999  Capital Metro HQ - East 5th at Broadway                55  

[8000 rows x 8 columns]

#####1. Berapa banyak label unique dari bike_type?

In [3]:
# working area
df_1 = client.query('''
SELECT COUNT(DISTINCT bike_type) AS total_bike_type
FROM `hacktiv8-ftds-challenge.practice.bike_rent`
''').to_dataframe()

df_1

total_bike_type
0                2

Insight: Ada dua macam sepeda yang direntalkan oleh layanan Austin Bikeshare

#####2. Apa saja nama label unique dari tipe sepeda?

In [4]:
# working area
df_2 = client.query('''
SELECT DISTINCT bike_type AS list_of_bike_type
FROM `hacktiv8-ftds-challenge.practice.bike_rent`
''').to_dataframe()

df_2

list_of_bike_type
0           classic
1          electric

Insight: Ada dua macam sepeda yang direntalkan, yakni sepeda tipe klasik, dan sepeda tipe elektrik

#####3. Berapa banyak total peminjaman sepeda pada tahun 2016 dan 2017?

In [5]:
# working area
df_3 = client.query('''
SELECT COUNT(start_time)
FROM `hacktiv8-ftds-challenge.practice.bike_rent`
WHERE EXTRACT(YEAR FROM start_time) = 2016
OR EXTRACT(YEAR FROM start_time) = 2017
''').to_dataframe()

df_3

# df3_2016_2017 = df_3[(df_3['start_time'].dt.year == 2016) | (df_3['start_time'].dt.year == 2017)]

# df3_2016_2017.start_time.count()

f0_
0  7993

Insight: ada 7993 kali peminjaman sepeda pada tahun 2016 dan 2017

#####4. Dimana station yang paling banyak dikunjungi sebagai titik keberangkatan pada tahun 2016?

In [6]:
# working area
df_4 = client.query('''
SELECT start_station_name,
COUNT(start_station_name) AS total
FROM `hacktiv8-ftds-challenge.practice.bike_rent`
WHERE EXTRACT(YEAR FROM start_time) = 2016
GROUP BY start_station_name
ORDER BY total DESC
LIMIT 1;
''').to_dataframe()

df_4

start_station_name  total
0  Bullock Museum @ Congress & MLK    394

Insight: Bullock Museum @ Congress & MLK merupakan stasiun keberangkatan yang paling sering dikunjungi oleh pengguna sewa sepeda pada 2016, dengan total kunjungan sebanyak 394 kali.

#####5. Berapa nilai minimum, maximum, dan rata-rata dari durasi perjalanan pada keseluruhan data?

In [7]:
# working area
df_5 = client.query('''
SELECT MIN(duration_minutes) AS minimum_trip_duration,
  MAX(duration_minutes) AS maximum_trip_duration,
  AVG(duration_minutes) AS mean_trip_duration
FROM `hacktiv8-ftds-challenge.practice.bike_rent`
''').to_dataframe()

df_5

minimum_trip_duration  maximum_trip_duration  mean_trip_duration
0                      2                   2218             35.0995

Insight: Rental sepeda paling singkat pada keseluruhan rekaman data ialah 2 menit. Rental sepeda paling lama ialah selama 2218 menit. Rata-rata durasi rental sepeda secara keseluruhan ialah 35 menit.

#####6. Berapa rata-rata durasi perjalanan dari subscriber_type `Student Membership` dan `Local31`?

In [8]:
# working area
df_6 = client.query('''
SELECT AVG(duration_minutes)
FROM `hacktiv8-ftds-challenge.practice.bike_rent`
WHERE subscriber_type = 'Student Membership' OR subscriber_type = 'Local31';
''').to_dataframe()

df_6

f0_
0  NaN

Insight: Tidak ada rata-rata lama durasi peminjaman sepeda oleh mereka yang memegang tipe subscription 'Student Membership' dan 'Local31'.

#####7. Tampilkan entry data yang memiliki `duration_time` diatas 4.25 jam. (*Gunakan subquery untuk menjawab pertanyaan ini dengan meng-convert duration_time dari menit ke hour*).

In [9]:
# working area
df_6 = client.query('''
SELECT duration_minutes
FROM `hacktiv8-ftds-challenge.practice.bike_rent`
WHERE duration_minutes/60 > 4.25
# WHERE 4.25 < (
#   SELECT duration_minutes/60
#   FROM `bigquery-public-data.austin_bikeshare.bikeshare_trips`

''').to_dataframe()
df_6

duration_minutes
0                351
1                350
2                783
3                789
4                325
..               ...
86               350
87               643
88               684
89               292
90               757

[91 rows x 1 columns]

Insight: Ada 91 record dengan durasi peminjaman lebih dari 4,25 jam

###**Pandas**

#####Data Loading

In [10]:
# working area
df

trip_id subscriber_type bike_id bike_type                start_time  \
0     18896589        Explorer      93   classic 2018-10-28 14:16:14+00:00   
1     13948351        Local365     315   classic 2017-03-07 12:37:33+00:00   
2     16243803        Local365     877   classic 2017-10-21 09:49:48+00:00   
3     16222446         Walk Up     273   classic 2017-10-18 15:01:52+00:00   
4     16222451         Walk Up     201   classic 2017-10-18 15:02:26+00:00   
...        ...             ...     ...       ...                       ...   
7995  15469480         Walk Up     327   classic 2017-07-22 20:17:46+00:00   
7996  16130650         Walk Up     528   classic 2017-10-07 11:25:36+00:00   
7997  14204138         Walk Up     235   classic 2017-03-27 13:19:02+00:00   
7998  15037809         Walk Up     206   classic 2017-06-05 11:42:40+00:00   
7999   9499076         Walk Up     315   classic 2016-04-09 18:52:09+00:00   

                           start_station_name  \
0                     Dean Keeton & Speedway    
1                                 Main Office   
2           Republic Square @ 5th & Guadalupe   
3             Bullock Museum @ Congress & MLK   
4             Bullock Museum @ Congress & MLK   
...                                       ...   
7995          Bullock Museum @ Congress & MLK   
7996                  Congress & Cesar Chavez   
7997          State Capitol @ 14th & Colorado   
7998          State Capitol @ 14th & Colorado   
7999  Capital Metro HQ - East 5th at Broadway   

                             end_station_name  duration_minutes  
0                              11th & Salina                178  
1                                 Main Office               198  
2                                 Main Office                60  
3                                 Main Office                60  
4                                 Main Office                60  
...                                       ...               ...  
7995          Bullock Museum @ Congress & MLK                55  
7996          Bullock Museum @ Congress & MLK                55  
7997          State Capitol @ 14th & Colorado                55  
7998          State Capitol @ 14th & Colorado                55  
7999  Capital Metro HQ - East 5th at Broadway                55  

[8000 rows x 8 columns]

Insight: Tabel ini merupakan dataset yang memuat record peminjaman sepeda yang mana terdiri dari kolom trip_id, subscriber_type, bike_id, bike_type, start_time, start_station_name, end_station_name, duration_name

#####1. Cek kesesuaian tipe data dan keberadaan missing values dengan `.info()`

In [11]:
# working area
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   trip_id             8000 non-null   Int64              
 1   subscriber_type     8000 non-null   object             
 2   bike_id             8000 non-null   object             
 3   bike_type           8000 non-null   object             
 4   start_time          8000 non-null   datetime64[ns, UTC]
 5   start_station_name  8000 non-null   object             
 6   end_station_name    8000 non-null   object             
 7   duration_minutes    8000 non-null   Int64              
dtypes: Int64(2), datetime64[ns, UTC](1), object(5)
memory usage: 515.8+ KB


Insight: Ada delapan kolom pada dataframe df (trip_id, subscriber_type, bike_id, bike_type, start_time, start_station_time, end_station_time, dan duration_minutes). Semua data dalam seluruh kolom tidak ada yang null. Dua kolom (trip_id dan duration_minutes) merupakan kolom yang memiliki data tipe integer/int64. Kolom start_station_time memiliki data tipe datetime64. Kolom subscriber_type, bike_id, bike_type, start_station_time, dan end_station_time merupakan kolom dengan tipe data string.

#####2. Buat kolom baru bernama `quarter`.
  
  Lakukan ekstraksi waktu (quarter dalam tahun) berdasarkan kolom `start_time` dengan ketentuan berikut:

<div align="center">

  |Keterangan Waktu|Bulan|
  |--|--|
  |Q1|Januari, Februari, Maret|
  |Q2|April, Mei, Juni|
  |Q3|Juli, Agustus, September|
  |Q4|Oktober, November, Desember|

</div>

>**Notes:**
>- Dalam menjawab soal Pandas no. 2, **tidak boleh** menggunakan ekstraksi waktu `Series.dt.quarter`.
- Lakukan `looping` dan `conditional` dengan memanfaatkan ekstraksi bulan `Series.dt.month` terhadap kolom `start_time`, pastikan tipe datanya ialah `datetime` agar dapat dilakukan ekstraksi.
- Simpan hasil looping ke dalam suatu variabel, dan buat kolom baru berisikan variabel sebelumnya.

In [12]:
# working area
# cleaning datanya dulu sebelum jawab nomor 2

quarterColumn = []

dfWithColumnQuarter = df.copy()

for entry in dfWithColumnQuarter.start_time.dt.month:
  if entry <= 3:
    quarterColumn.append('Q1')
  elif entry <= 6:
    quarterColumn.append('Q2')
  elif entry <= 9:
    quarterColumn.append('Q3')
  else:
    quarterColumn.append('Q4')

# dfWithColumnQuarter['quarter'] = dfWithColumnQuarter.start_time.dt.month

dfWithColumnQuarter['quarter'] = quarterColumn

dfWithColumnQuarter

trip_id subscriber_type bike_id bike_type                start_time  \
0     18896589        Explorer      93   classic 2018-10-28 14:16:14+00:00   
1     13948351        Local365     315   classic 2017-03-07 12:37:33+00:00   
2     16243803        Local365     877   classic 2017-10-21 09:49:48+00:00   
3     16222446         Walk Up     273   classic 2017-10-18 15:01:52+00:00   
4     16222451         Walk Up     201   classic 2017-10-18 15:02:26+00:00   
...        ...             ...     ...       ...                       ...   
7995  15469480         Walk Up     327   classic 2017-07-22 20:17:46+00:00   
7996  16130650         Walk Up     528   classic 2017-10-07 11:25:36+00:00   
7997  14204138         Walk Up     235   classic 2017-03-27 13:19:02+00:00   
7998  15037809         Walk Up     206   classic 2017-06-05 11:42:40+00:00   
7999   9499076         Walk Up     315   classic 2016-04-09 18:52:09+00:00   

                           start_station_name  \
0                     Dean Keeton & Speedway    
1                                 Main Office   
2           Republic Square @ 5th & Guadalupe   
3             Bullock Museum @ Congress & MLK   
4             Bullock Museum @ Congress & MLK   
...                                       ...   
7995          Bullock Museum @ Congress & MLK   
7996                  Congress & Cesar Chavez   
7997          State Capitol @ 14th & Colorado   
7998          State Capitol @ 14th & Colorado   
7999  Capital Metro HQ - East 5th at Broadway   

                             end_station_name  duration_minutes quarter  
0                              11th & Salina                178      Q4  
1                                 Main Office               198      Q1  
2                                 Main Office                60      Q4  
3                                 Main Office                60      Q4  
4                                 Main Office                60      Q4  
...                                       ...               ...     ...  
7995          Bullock Museum @ Congress & MLK                55      Q3  
7996          Bullock Museum @ Congress & MLK                55      Q4  
7997          State Capitol @ 14th & Colorado                55      Q1  
7998          State Capitol @ 14th & Colorado                55      Q2  
7999  Capital Metro HQ - East 5th at Broadway                55      Q2  

[8000 rows x 9 columns]

#####3. Tampilkan data perjalanan dari tipe member `Local31` dengan durasi perjalanan diatas 1 jam.
*(Jika tidak ada entry yang memenuhi kondisi ini tidak apa, tuliskan kode pandasnya dengan lengkap saja)*

In [13]:
# working area

dfLocal31 = df.copy()

dfLocal31 = dfLocal31[(dfLocal31['subscriber_type'] == 'Local31') & (dfLocal31['duration_minutes'] > 60)]

dfLocal31.sort_values(by='duration_minutes')

Empty DataFrame
Columns: [trip_id, subscriber_type, bike_id, bike_type, start_time, start_station_name, end_station_name, duration_minutes]
Index: []

Insight: Tidak ada data yang memuat subscriber_type 'Local31' dengan data durasi perjalanan lebih lama dari satu jam.

#####4. Berapa rata-rata durasi perjalanan dari tipe pelanggan `Walk Up` dengan tipe sepeda `classic` pada bulan `Maret` dari data yang diambil?
*(Jika tidak ada entry yang memenuhi kondisi ini tidak apa, tuliskan kode pandasnya dengan lengkap saja)*

In [35]:
# working area

dfWalkUpClassicinMarch = df.copy()

dfWalkUpClassicinMarch = dfWalkUpClassicinMarch[(dfWalkUpClassicinMarch['subscriber_type'] == 'Walk Up') & (dfWalkUpClassicinMarch['bike_type'] == 'classic') & (dfWalkUpClassicinMarch.start_time.dt.month == 3)]

dfWalkUpClassicinMarch.duration_minutes.mean()


38.476584022038566

Insight: Rata-rata durasi perjalanan dari tipe pelanggan 'Walk Up' dengan tipe sepeda 'classic' khususnya yang muncul setiap bulan Maret ialah 38 menit.